src/Libraries/Microsoft.Extensions.AI.Evaluation.NLP/BLEU/BLEUAlgorithm.cs

```diff
@@ -0,0 +1,205 @@
+        }
+    }
+
+    internal static int ClosestRefLength(IEnumerable<IEnumerable<string>> references, int hypLength)
+    {
+        if (!references.Any())
+        {
+            return 0;
+        }
+
+        int closestRefLength = 0;
+        int closest = int.MaxValue;
+        foreach (var reference in references)
+        {
+            int refLength = reference.Count();
+            int diff = System.Math.Abs(refLength - hypLength);
+            if (diff < closest ||
+               (diff == closest && refLength < closestRefLength))
+            {
+                closest = diff;
+                closestRefLength = refLength;
+            }
+        }
+
+        return closestRefLength;
+    }
+
+    internal static double BrevityPenalty(int closestRefLength, int hypLength)
+    {
+        if (hypLength <= 0)
+        {
+            return 0.0;
+        }
+
+        if (closestRefLength <= 0 || hypLength > closestRefLength)
+        {
+            return 1.0;
+        }
+
+        return System.Math.Exp(1 - ((double)closestRefLength / hypLength));
+    }
+
+    internal static RationalNumber ModifiedPrecision(IEnumerable<IEnumerable<string>> references, IEnumerable<string> hypothesis, int n = 1)
+    {
+        if (n <= 0)
+        {
+            Throw.ArgumentOutOfRangeException(nameof(n), "N must be greater than zero.");
+        }
+
+        if (!references.Any() || !hypothesis.Any())
+        {
+            return new RationalNumber(0, 0);
+        }
+
+        var hyp = NGrams(hypothesis, n).ToArray();
+        var hypCounts = new MatchCounter<NGram<string>>(hyp);
+
+        Dictionary<NGram<string>, int> maxCounts = [];
+
+        var matchCounts = new MatchCounter<NGram<string>>();
+
+        foreach (var rf in references)
+        {
+            var refGrams = NGrams(rf, n).ToArray();
+            var refCounts = new MatchCounter<NGram<string>>(refGrams);
+
+            foreach (var ct in refCounts.Values)
+            {
+                if (maxCounts.TryGetValue(ct.Key, out int val))
+                {
+                    maxCounts[ct.Key] = System.Math.Max(val, ct.Value);
+                }
+                else
+                {
+                    maxCounts[ct.Key] = ct.Value;
+                }
+            }
+        }
+
+        Dictionary<NGram<string>, int> clippedCounts = [];
+        foreach (var h in hypCounts.Values)
+        {
+            if (maxCounts.TryGetValue(h.Key, out var v))
+            {
+                clippedCounts[h.Key] = System.Math.Min(h.Value, v);
+            }
+            else
+            {
+                // If the hypothesis n-gram is not in any reference, it is clipped to 0.
+                clippedCounts[h.Key] = 0;
+            }
+        }
+
+        int numerator = clippedCounts.Values.Sum();
+        int denominator = System.Math.Max(1, hypCounts.Sum);
+
+        return new RationalNumber(numerator, denominator);
+    }
+
+    internal static double[] EqualWeights(int n)
```

Comment:
It is hard to understand what this function is for just by looking at the name. Consider renaming to `GetDefaultWeights` or `GetDefaiultBLEUWeights`...

Reply:
I've added comment documenting it. There is a property right below it named `DefaultBLEUWeights`, which is for that purpose.